runtime/parachains/src/scheduler.rs

```diff
@@ -13,3 +13,1403 @@
 
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The scheduler module for parachains and parathreads.
+//!
+//! This module is responsible for two main tasks:
+//!   - Paritioning validators into groups and assigning groups to parachains and parathreads
+//!   - Scheduling parachains and parathreads
+//!
+//! It aims to achieve these tasks with these goals in mind:
+//! - It should be possible to know at least a block ahead-of-time, ideally more,
+//!   which validators are going to be assigned to which parachains.
+//! - Parachains that have a candidate pending availability in this fork of the chain
+//!   should not be assigned.
+//! - Validator assignments should not be gameable. Malicious cartels should not be able to
+//!   manipulate the scheduler to assign themselves as desired.
+//! - High or close to optimal throughput of parachains and parathreads. Work among validator groups should be balanced.
+//!
+//! The Scheduler manages resource allocation using the concept of "Availability Cores".
+//! There will be one availability core for each parachain, and a fixed number of cores
+//! used for multiplexing parathreads. Validators will be partitioned into groups, with the same
+//! number of groups as availability cores. Validator groups will be assigned to different availability cores
+//! over time.
+
+use sp_std::prelude::*;
+use sp_std::convert::TryInto;
+use primitives::{
+	parachain::{Id as ParaId, CollatorId, ValidatorIndex},
+};
+use frame_support::{
+	decl_storage, decl_module, decl_error,
+	weights::Weight,
+};
+use codec::{Encode, Decode};
+use sp_runtime::traits::{Saturating, Zero};
+
+use rand::{SeedableRng, seq::SliceRandom};
+use rand_chacha::ChaCha20Rng;
+
+use crate::{configuration, paras, initializer::SessionChangeNotification};
+
+/// The unique (during session) index of a core.
+#[derive(Encode, Decode, Default, PartialOrd, Ord, Eq, PartialEq, Clone, Copy)]
+#[cfg_attr(test, derive(Debug))]
+pub struct CoreIndex(u32);
```

Comment:
Add a `Deref` impl so most of the `.0` can be omitted.

Reply:
That is considered bad practice as this is not a smart pointer. From https://doc.rust-lang.org/stable/std/ops/trait.Deref.html

> On the other hand, the rules regarding Deref and DerefMut were designed specifically to accommodate smart pointers. Because of this, Deref should only be implemented for smart pointers to avoid confusion.